test/helpers/restricted-nori.ts

```diff
@@ -0,0 +1,140 @@
+
+export const mintAndListRemovals = async ({
+  testSetup,
+  projectId = 1_234_567_890,
+  scheduleStartTime = NOW,
+  removalDataToList = [],
+}: {
+  testSetup: Awaited<ReturnType<typeof setupTest>>;
+  projectId?: number;
+  scheduleStartTime?: number;
+  removalDataToList?: {
+    amount: number;
+    vintage?: number;
+    supplier?: string;
+  }[];
+}): Promise<{
+  listedRemovalIds: BigNumber[];
+  projectId: number;
+  scheduleStartTime: number;
+}> => {
+  const { hre, removal, fifoMarket, rNori } = testSetup;
+  let tokenIds: BigNumber[] = [];
+  if (removalDataToList.length > 0) {
+    const { supplier } = hre.namedAccounts;
+    const defaultStartingVintage = 2016;
+    tokenIds = await Promise.all(
+      removalDataToList.map((removalData, index) => {
+        return createRemovalTokenId({
+          removal,
+          hre,
+          removalData: {
+            supplierAddress: removalData.supplier ?? supplier,
+            vintage: removalData.vintage ?? defaultStartingVintage + index,
+          },
+        });
+      })
+    );
+    const removalBalances = removalDataToList.map((removalData) =>
+      hre.ethers.utils.parseUnits(removalData.amount.toString())
+    );
+
+    const packedData = hre.ethers.utils.defaultAbiCoder.encode(
+      ['uint256', 'uint256', 'address', 'bool'],
+      [projectId, scheduleStartTime, fifoMarket.address, true]
+    );
+    expect(
+      await removal.mintBatch(supplier, removalBalances, tokenIds, packedData)
+    ).to.emit(rNori, 'ScheduleCreated');
+  }
+
+  return {
+    listedRemovalIds: tokenIds,
+    projectId,
+    scheduleStartTime,
+  };
+};
+
+export const formatTokensReceivedUserData = (removalId: BigNumber): any => {
+  return hre.ethers.utils.defaultAbiCoder.encode(['uint256'], [removalId]);
+};
+
+export const restrictRemovalProceeds = async ({
+  testSetup,
+  removalIds,
+  removalAmountsToRestrict,
+}: {
+  testSetup: Awaited<ReturnType<typeof setupTest>>;
+  removalIds: BigNumber[];
+  removalAmountsToRestrict: number[];
+}): Promise<void> => {
+  const { rNori, bpNori } = testSetup;
+  await Promise.all(
+    removalIds.map((id, index) => {
+      const userData = formatTokensReceivedUserData(id);
+      return bpNori.send(
+        rNori.address,
+        removalAmountsToRestrict[index],
+        userData
+      );
+    })
+  );
+};
+
+export const compareScheduleDetailForAddressStructs = (
+  receivedScheduleDetail: RestrictedNORI.ScheduleDetailForAddressStruct,
+  expectedScheduleDetail: Partial<RestrictedNORI.ScheduleDetailForAddressStruct>
+): void => {
+  const keys = [
+    'tokenHolder',
+    'scheduleTokenId',
+    'startTime',
+    'endTime',
+    'balance',
+    'claimableAmount',
+    'claimedAmount',
+    'quantityRevoked',
+    'exists',
+  ] as const;
+  for (const key of keys) {
+    if (key in expectedScheduleDetail) {
+      expect(receivedScheduleDetail[key]).to.equal(expectedScheduleDetail[key]);
```

Comment:
don't love this as it breaks [no-conditional-expect](https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-conditional-expect.md)

Reply:
hmm, can you think of a better way to achieve this without the conditional that still allows me to compare a complete received object to a partial expectation? (I want to be able to craft expectations that only use the few keys that I actually care about)

Reply:
like this? 

```
expect({ a: 1, b: 2 }).to.include({ b: 2 }); 
```

Alternative: https://www.chaijs.com/api/bdd/#method_property

Reply:
any luck here?

Reply:
no, i'm struggling because these objects that come back look like this:

Reply:
there's also the issue that the array `tokenHolders` needs to be compared using a `to.have.members` because comparisons are strict by default instead of deep

Reply:
I didn't even know you could have key:value pairs as array items, but taht's what seems to be the case for the received structs.  I've been trying to turn the expected object into something that could be compared as a subset of that incoming array but still not having any luck.  I understand the rationale behind no conditional expect but I feel like in this case we're not really creating the types of problems it's supposed to solve.  wondering if we can just not block this PR on this for now. really at a limit here

Reply:
You can't. But in JS it's objects all the way down so there's some wonky manipulation happening here. I really dislike the design decision on this one. Always forget that it's an issue when it comes to test and have always wanted a built in way to just return an object. 

Don't like that this is calling expect like this and it should at some point be extracted to something else (e.g., pretty sure you can slice the array starting at length/2+1 or something and spread it into an object this avoiding any looped/conditional assertion

Reply:
https://github.com/ethers-io/ethers.js/discussions/2429

Reply:
Also there is a deep variant of the above fyi